clippy_lints/src/unused_unit.rs

```diff
@@ -159,18 +159,18 @@ fn get_def(span: Span) -> Option<Span> {
 }
 
 fn lint_unneeded_unit_return(cx: &LateContext<'_>, ty_span: Span, span: Span) {
-    let (ret_span, appl) =
-        span.with_hi(ty_span.hi())
-            .get_source_text(cx)
-            .map_or((ty_span, Applicability::MaybeIncorrect), |src| {
-                position_before_rarrow(&src).map_or((ty_span, Applicability::MaybeIncorrect), |rpos| {
-                    (
-                        #[expect(clippy::cast_possible_truncation)]
-                        ty_span.with_lo(BytePos(span.lo().0 + rpos as u32)),
-                        Applicability::MachineApplicable,
-                    )
-                })
-            });
+    let (ret_span, appl) = if let Some(src) = span.with_hi(ty_span.hi()).get_source_text(cx)
+        && let Some(rpos) = position_before_rarrow(&src)
+    {
+        (
+            #[expect(clippy::cast_possible_truncation)]
+            ty_span.with_lo(BytePos(span.lo().0 + rpos as u32)),
+            Applicability::MachineApplicable,
+        )
+    } else {
+        (ty_span, Applicability::MaybeIncorrect)
+    };
```

Comment:
Further cleanups (to squash into the first commit):

```suggestion
    let (ret_span, appl) =
        if let Some(Some(rpos)) = span.with_hi(ty_span.hi()).with_source_text(cx, position_before_rarrow) {
            (
                ty_span.with_lo(span.lo() + BytePos::from_usize(rpos)),
                Applicability::MachineApplicable,
            )
        } else {
            (ty_span, Applicability::MaybeIncorrect)
        };
```

`with_source_text()` avoid building a `SourceText` object. You'll need to import `rustc_span::Pos as _` to get `BytePos::from_usize()`.